tests/manual_download_data_pthread.html

```diff
@@ -0,0 +1,198 @@
+          this.totalDependencies = Math.max(this.totalDependencies, left);
+          Module.setStatus(left ? 'Preparing... (' + (this.totalDependencies-left) + '/' + this.totalDependencies + ')' : 'All downloads complete.');
+        }
+      };
+      Module.setStatus('Downloading...');
+      window.onerror = function() {
+        Module.setStatus('Exception thrown, see JavaScript console');
+        spinnerElement.style.display = 'none';
+        Module.setStatus = function(text) {
+          if (text) Module.printErr('[post-exception status] ' + text);
+        };
+      };
+
+      function download(url) {
+        return new Promise(function(resolve, reject) {
+          var xhr = new XMLHttpRequest();
+          xhr.open('GET', url, true);
+          xhr.responseType = 'arraybuffer';
+          xhr.onload = function() { resolve(xhr.response); };
+          xhr.onerror = function(e) { reject(e); };
+          xhr.send(null);
+        });
+      }
+
+      function addScriptToDom(scriptCode) {
+        return new Promise(function(resolve, reject) {
+          var script = document.createElement('script');
+          var blob = new Blob([scriptCode], { type: 'application/javascript' });
+          var objectUrl = URL.createObjectURL(blob);
+          script.src = objectUrl;
+          script.onload = function() {
+            console.log('added js script to dom');
+            script.onload = script.onerror = null; // Remove these onload and onerror handlers, because these capture the inputs to the Promise and the input function, which would leak a lot of memory!
+            URL.revokeObjectURL(objectUrl); // Free up the blob. Note that for debugging purposes, this can be useful to comment out to be able to read the sources in debugger.
+            resolve();
+          }
+          script.onerror = function(e) {
+            script.onload = script.onerror = null; // Remove these onload and onerror handlers, because these capture the inputs to the Promise and the input function, which would leak a lot of memory!
+            URL.revokeObjectURL(objectUrl);
+            console.error('script failed to add to dom: ' + e);
+            reject(e.message || "(out of memory?)");
+          }
+          document.body.appendChild(script);
+        });
+      }
+
+      Module.getPreloadedPackage = function(remotePackageName, remotePackageSize) {
+        console.log('Runtime asking for remote package ' + remotePackageName + ', expected size ' + remotePackageSize + 'bytes.');
+        return Module['downloadedData'];
+      }
+
+      var dataDownload = download('/test/manual_download_data.data').then(function(data) {
+        console.log('downloaded data file');
+        Module['downloadedData'] = data;
+        var jsDownload = download('manual_download_data.js').then(function(data) {
+          Module['mainScriptUrlOrBlob'] = new Blob([data]);
+          console.log('downloaded js file');
+          addScriptToDom(data);
+        });
```

Comment:
Adding this entire file makes it hard to see what is being tested.

Can we not just take this two JS functions and put them a in `--pre-js` files?   

Also, would it be enough to just do this in a pre-js?

```
Module.getPreloadedPackage = function(remotePackageName, remotePackageSize) {
  console.log('getPreloadedPackage called')
  assert(!ENVIRONMENT_IS_PHTREAD, "getPreloadedPackage should only ever be called from the main thread");
  // We can't/don't actually do any pre-loading, we are just checked that we are never called from a thread.
  return false;
}
```

I think it we can simplify this test to just this we don't need to also add a test to test_core.py.

Reply:
Overloading `Module.getPreloadedPackage` will still fail because when the generated worker file loads the main glue file, it doesn't set `Module.getPreloadedPackage`. That is why the glue file tries to access  `/MyLibrary.data` from server root. Modifying worker.js to set `Module.getPreloadedPackage` was a part of my very first proposed solution/commit, but it is unnecessary given that file system is being proxied.

Re: just using `--pre-js` files, I'm not sure if I can use that to test my exact error before these changes - which is that the preload phase fails when the .data file is not at server root, while I we are replicating that exact scenario in this test case. Do you see a way to test it accurately with just pre-js files?

Reply:
Another point in favor of current test is that the existing `test_preload_file_with_manual_data_download` was never checking for pthreads, which is why the bug wasn't caught before. So IMO we should have an equivalent test with pthreads enabled. I can combine the two tests to use the same .html file instead of creating a new `tests/manual_download_data_pthread.html` if that helps reduce the noise.